spec/p2p/reactor/reactor.qnt

```diff
@@ -0,0 +1,238 @@
+/*
+ * Reactor is responsible for handling incoming messages on one or more
+ * Channel. Switch calls GetChannels when reactor is added to it. When a new
+ * peer joins our node, InitPeer and AddPeer are called. RemovePeer is called
+ * when the peer is stopped. Receive is called when a message is received on a
+ * channel associated with this reactor.
+ */
+// Code: https://github.com/cometbft/cometbft/blob/main/p2p/base_reactor.go
+module reactor {
+
+	// Unique ID of a node.
+	type NodeID = str
+
+	/*
+	 * Peer is an interface representing a peer connected on a reactor.
+	 */
+	type Peer = {
+		ID: NodeID,
+
+		// Other fields can be added to represent the p2p operation.
+	}
+
+	// Byte ID used by channels.
+	type Byte = int
+
+	// Channel configuration.
+	// Only the Byte ID is relevant here.
+	type ChannelDescriptor = {
+		ID: Byte,
+		Priority: int,
+	}
+
+	/*
+	 * Envelope contains a message with sender routing info.
+	 */
+	type Envelope = {
+		Src: Peer, // Sender
+		Message: str, // Payload
+		ChannelID: Byte,
+	}
+
+	// A Routine is used to interact with an active Peer.
+	type Routine = {
+		name: str,
+		peer: Peer,
+	}
+
+	type ReactorState = {
+		// Peers that have been initialized but not yet removed.
+		// The reactor should expect receiving messages from them.
+		peers: Set[NodeID],
+
+		// The reactor runs multiple routines.
+		routines: Set[Routine],
+
+		// Service implementation.
+		started: bool,
```

Comment:
That is a good suggestion. Here I was blindly following the implementation (Go code). They use booleans here so that to use atomic read-compare-write which are very efficient, but renders the code hard to read.